### PR TITLE
Add std.stdio.File.setRawMode().

### PR DESCRIPTION
**Pull request by JohanEngelen**
Especially for files that cannot be opened by the user in binary mode (like stdin and stdout), having only rawWrite and rawRead available is very limiting. To enable binary/raw file operation for stdin and stdout, the code to set those streams to binary mode is non-trivial. setRawMode() resolves this issue.

### CODE REVIEW

**Comment by JohanEngelen** [comment]
(made fixes, squashed into one commit again)

**Comment by Hackerpilot** [comment]
Looks reasonable.

**Comment by quickfur** [comment]
Hmm. I don't like exposing OS-level implementation details like this. I prefer `lockingBinaryWriter` as implemented in PR https://github.com/D-Programming-Language/phobos/pull/2011. Under the hood it basically does the same thing, but the API is cleaner, works nicely as an output range, and fits in better with the already existing `lockingTextWriter`.

**Comment by CyberShadow** [comment]
I don't disagree with the idea, but this needs to be a toggle. "Binary mode" is also the correct term here.

An alternative would be to provide a way to reopen stdin/out/err with arbitrary modes.

> I don't like exposing OS-level implementation details like this.

For the record this is the C API and Windows-only, so nothing OS-level.

BTW, this needs a correspoding Bugzilla issue.

**Comment by JohanEngelen** [comment]
Related bug reports:
https://issues.dlang.org/show_bug.cgi?id=7033
https://issues.dlang.org/show_bug.cgi?id=4243
https://issues.dlang.org/show_bug.cgi?id=9455
https://issues.dlang.org/show_bug.cgi?id=10545

**Comment by CyberShadow** [comment]
> https://issues.dlang.org/show_bug.cgi?id=9455
> https://issues.dlang.org/show_bug.cgi?id=10545

Perhaps these can be combined into a property, `binaryMode`. This would allow querying and toggling either way.

**Comment by JohanEngelen** [comment]
About lockingBinaryWriter: yes, would be great! But, it is not exactly the same:  it locks.

About OS-level stuff: I think the basic problem is not setmode, it is the Digital Mars IO stuff that is troublesome. There should be a function (with sane name) for `atomicOp!"&="(__fhnd_info[fd], ~FHND_TEXT);`.

I first thought about setBinaryMode, setTextMode, but then I thought that that would somehow expose to much detail (and it doesn't actually set binary mode on non-Windows). Then I figured that the functionality is basically a permanent rawWrite, so "raw" stuck.
Anyway, I am happy to change the PR naming / toggling to your liking. However, I don't want to keep resubmitting this thing; so please discuss and come to a conclusion for me :-)

File: std/stdio.d [hunk 1]
```diff
@@ -903,6 +903,56 @@ Throws: $(D ErrnoException) if the file is not opened or if the call to $(D fwri
...
+    {
+        flush(); // before changing translation mode
+        version(Windows)
+        {
+            immutable fd = ._fileno(_p.handle);
+            ._setmode(fd, _O_BINARY);
+            version(DIGITAL_MARS_STDIO)
+            {
+                import core.atomic;
+                immutable info = __fhnd_info[fd];
```
Comment (JohanEngelen) [inline std/stdio.d]: variable not used, remove.

**Comment by JakobOvrum** [comment]
Note that if you include `Fix ticket 7033` etc. in the commit message, a bot will automatically link them and mark them as resolved when/if the commits are merged.

File: std/stdio.d [hunk 2]
```diff
@@ -903,6 +903,56 @@ Throws: $(D ErrnoException) if the file is not opened or if the call to $(D fwri
...
+Calls $(LREF flush) (all platforms), and sets the $(D FILE*) to binary mode on Windows.
+
+This can be used to set $(D stdin) and $(D stdout), which are implicitly opened in text
+mode, to binary mode on Windows.
+
+See_Also:
+$(LREF rawRead)
+$(LREF rawWrite)
+ */
+    void setRawMode()
```
Comment (andralex) [inline std/stdio.d]: The terminology is binary vs. text. Then we probably need a means to get the current mode, so there's no escaping the API:

```d
enum FileMode { text, binary }
void getMode(FileMode);
void setMode(FileMode);
```

File: std/stdio.d [hunk 3]
```diff
@@ -903,6 +903,56 @@ Throws: $(D ErrnoException) if the file is not opened or if the call to $(D fwri
...
+    {
+        flush(); // before changing translation mode
```
Comment (andralex) [inline std/stdio.d]: I hope this throws on error.

**Comment by andralex** [comment]
ping?

**Comment by JohanEngelen** [comment]
sorry, guess I should have closed this.

**Pull request closed without merging**